torch/nn/parallel/distributed.py

```diff
@@ -612,6 +625,172 @@ def train(self, mode=True):
+          >>>      model = nn.Linear(1, 1, bias=False).to(rank)
+          >>>      model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[rank], output_device=rank)
+          >>>      # Rank 1 gets one more input thank rank 0.
+          >>>      inputs = [torch.tensor([1]).float() for _ in range(10 + rank)]
+          >>>      with model.join():
+          >>>          for _ in range(5):
+          >>>              for inp in inputs:
+          >>>                  loss = model(inp).sum()
+          >>>                  loss.backward()
+          >>>  torch.cuda.synchronize(device=rank)
+        """
+        try:
+            has_error = False
+            self.ddp_join_enabled = enable
+            yield
+        except Exception as e:
+            # Set to skip any processing in the finally block.
+            has_error = True
+            raise e
+        finally:
+            # Skip any processing to let the exception immediately be raised if
+            # there was one.
+            if enable and not has_error:
+                all_procs_joined = False
+                is_last_joiner = True
+                buckets_rebuilt = False
+                # Schedule an allreduce to match fwd pass allreduce in non-joined procs
+                while not all_procs_joined:
+                    num_active_procs = self._schedule_shadow_all_reduce_for_fwd_pass()
+                    if num_active_procs == 0:
+                        all_procs_joined = True
+                    else:
+                        # Some DDP process still needs to be joined.
+                        if is_last_joiner:
+                            is_last_joiner = False
+                        # Schedule a corresponding broadcast if we are syncing module
+                        # buffers in the forward pass.
+                        self._check_and_sync_module_buffers()
+                        # Schedule an allreduce to match the backwards pass allreduce.
+                        self._match_all_reduce_for_bwd_pass()
+                        # Check if we need to allreduce locally unused params.
+                        if self.find_unused_parameters:
+                            self._match_unused_params_allreduce()
+                        # Check if we need to rebuild buckets to match broadcast calls
+                        # from other processes.
+                        if not self.find_unused_parameters and not buckets_rebuilt:
+                            # Does no work if buckets are already rebuilt.
+                            self.reducer._push_all_rebuilt_params()
+                            # If buckets not rebuilt, will rebuild and
+                            # re-intialize tensor/bucket mapping so that future
+                            # allreduce orders are consisted across joined and
+                            # un-joined processes.
+                            self.reducer._rebuild_buckets()
```

Comment:
Would I be correct if I assume it is OK even if the bucket has already been rebuilt as the `_rebuild_buckets` will skip it?